Blogpost about docviewer updates (#474)
SenseNet/sensenet.github.io: _posts/2021-01-15-docviewer-updates.md

---
title: "Insight to SNaaS delivery process"
author: [pusztaienike]
image: "../img/posts/docviewer_refactor.jpg"
tags: [refactor, document viewer, saas, docviewer, document]
redirect_to: https://www.sensenet.com/blog/2020-01-15-docviewer_updates
---

🌇 “Cities grow, cities evolve, cities have parts that simply die while other parts flourish; each city has to be renewed in order to meet the needs of its populace…” 🌆 -Grady Booch

---

The basic purpose of code refactoring is to make the code more efficient and maintainable. Of course we are about to keep our code up-to-date and optimize its performance so we thought that it is time to clean a lot of complicated lines of code in the document viewer. ♻️

# Logic ✏️✂️🚀🐛👭

> The main goal was to simplify the code of the document viewer. There were some magic numbers, unnecessary props, useEffects that runs many times, so we had to go deep to rethink the logic. After planning, we had to rebuild the function based on the existing code. Although we try to avoid code repetition, sometimes it is even necessary to break down existing logic according to its function. As a result, Thumbnails and PageList became two separate components, for example. ✏️

Reviewer: ```suggestion
The main goal was to simplify the code. There were some magic numbers, unnecessary props, useEffects that run many times, so we had to go deep to rethink the logic. After planning, we had to rebuild functions based on the existing code. Although we try to avoid code repetition, sometimes it is even necessary to break down existing logic according to its function. As a result, Thumbnails and PageList became two separate components, for example. ✏️
```

> The main goal was to simplify the code of the document viewer. There were some magic numbers, unnecessary props, useEffects that runs many times, so we had to go deep to rethink the logic. After planning we had to try to rebuild the function based on the existing code. Although we basically try to avoid code repetition, sometimes it is even necessary to break down existing logic according to its function. As a result, Thumbnails and PageList became two separate components for example. ✏️

Reviewer 2: ```suggestion
The main goal was to simplify the code of the document viewer. There were some magic numbers, unnecessary props, useEffects that runs many times, so we had to go deep to rethink the logic. After planning, we had to rebuild the function based on the existing code. Although we try to avoid code repetition, sometimes it is even necessary to break down existing logic according to its function. As a result, Thumbnails and PageList became two separate components, for example. ✏️
```

Reviewer 2: "to break down existing logic according to its function."
Nem pontosan értem ezt a részt, leírod kérlek, hogy mit jelent?

Reviewer 3: Story: Volt egy komponens, aminek az volt a neve, hogy PageList. Ez a komponens látta el  a feladatokat a docviewer fő képernyőjén(ahol a preview képek között lehet görgetni) és a thumnails-eknél is (az oldalak lekicsinyített mása). Sajnos rengeteg problémánk adódott abból, hogy ezen a két helyen ugyanazt a komponenst használtuk, mert bár kinézetre nagyon ugyanúgy néztek ki, a funkciójukban teljesen eltérnek. Ezért határoztam úgy, hogy ha esetleg kódismétlés is lesz, ha szétválasztom őket, mert elég hasonló és a logika egy részét lehet kétszer kell majd leírni akkor is jobb, mint ha rosszul működne. A mondat valahogy úgy nézne ki, hogy: "Bár alapból igyekszünk elkerülni a kódismétlést, néha fontosabb, hogy elválasszuk egymástól a meglévő logikát a funkciójuk alapján"

> Not only the code has become simpler but also features. For example, rotating pages is only a form of visual representation, so it is unnecessary to store this information, it can be reproduced anytime. SearchBar widget become redundant too, because all browsers can search on page, so we removed this from selectable widgets. ✂️

Reviewer: ```suggestion
Not only the code has become simpler but also some of the features. For example, rotating pages is only a form of visual representation, so it is unnecessary to store this information, it can be reproduced anytime. SearchBar widget become redundant too, because all browsers can search on page, so we removed this from selectable widgets. ✂️
```

> Not only the code has become simpler but also features. For example, rotating pages is only a form of visual representation, so it is unnecessary to store this information, it can be reproduced anytime. SearchBar widget become redundant too, because all browsers can search on page, so we removed this from selectable widgets. ✂️

Reviewer 4: "but also features" - I miss "the", "but also the features"

> Not only has the code become simpler but also features as well. For example rotating pages is only a form of visual representation, so it is unnecessary to store this information, it can be reproduced at any time. SearchBar widget become redundant too, because all browsers can search on page, so we removed this from selectable widgets. ✂️

Reviewer 2: ```suggestion
Not only the code has become simpler but also features. For example, rotating pages is only a form of visual representation, so it is unnecessary to store this information, it can be reproduced anytime. SearchBar widget become redundant too, because all browsers can search on page, so we removed this from selectable widgets. ✂️
```

> We also put a lot of emphasis on improving performance as well. Sometimes we can also achieve significant results by making small changes such as taking advantage of the opportunities provided by our framework: replacing useEffect with [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo). 🚀

Reviewer 2: ```suggestion
We've put a great emphasis on improving performance as well. Sometimes we can achieve significant results by making small changes, such as taking advantage of the opportunities provided by our framework: replacing useEffect with [useMemo](https://reactjs.org/docs/hooks-reference.html#usememo). 🚀
```

> Which is also important during these kind of full review is finding and fixing bugs in the software. Of course, during the re-rolling of this function, we also encountered many cases where a different solution had to be applied than we originally invented. 🐛

Reviewer 2: ```suggestion
Finding and fixing bugs in the software is also important in case of a full review like this. Of course, during the re-rolling of this function, we also encountered many cases where a different solution had to be applied than we originally planned. 🐛
```

The point is that we tried to look at the possibilities from the user's point of view and we were trying to make this feature user friendly.👭

# Style 🎨

> We are also spreading another new option to users, which is overwriting styles from a higher level. It is very important that the user who uses our service can shape it for her/his own needs. So far there have been relatively few opportunities for it, but we are changing that now:

Reviewer 2: ```suggestion
We will release another new option to users, which is overwriting styles from a higher level. It is very important to enable users to shape our solution for their own needs. So far there have been relatively few opportunities for it, but we are changing that now:
```

> You can define you style on your own and you can pass it to the component itself, like:

Reviewer 4: ```suggestion
You can define your style on your own and you can pass it to the component itself, like:
```

```javascript
const useDisabledStyle = makeStyles(() =>
  createStyles({
    iconButton: {
      "&:disabled": {
        opacity: 0.26,
      },
    },
  })
);

const disabledStyle = useDisabledStyle();
```

If you use typescript you can also see the possible classes what you can overwrite:

<p align="center">
<img src="/img/posts/docviewer/overwrite_styles.png">
</p>

> Our designer is constantly coming up with new solutions, the style is constantly changing, we can never consider it ready. 🎡

Reviewer 2: ```suggestion
Since our designer always come up with new solutions and the style is constantly changing, we can never consider it done. 🎡
```

Reviewer 2: azért jobb ide a done szerintem mert használatra kész van, csak "befejezve" nem lesz soha.

# New features 🎉

For those who have been using sensenet for a long time, it may not be new, but if you have just joined, you may come across the keywords redaction, highlight and annotation for the first time.

**Redaction** 🔳

> Sometimes when you are working with sensitive information 🗝️ in documents you don’t want to share them with everyone, just who has the appropriate privileges. But what if you would like to share the other parts of the document? This is the case when you need to apply redaction.

Reviewer 4: What if you would like to share the document except the sensitive information? - I would say..

> Sometimes when you are working with sensitive information 🗝️ in documents you don’t want to share them with everyone, just who has the appropriate privileges. But what if you would like to share the other parts of the document? This is the case when you need to apply redaction.

Reviewer 2: ```suggestion
Sometimes when you are working with sensitive information 🗝️ in documents you don’t want to share them with everyone, only users having the appropriate privileges. But what if you would like to share the other parts of the document? This is the case when you need to apply redaction.
```

> With this feature you can hide parts of your doc with a black rectangle. If another user would like to see it, it is only possible if she/he has the permission for it, otherwise the page will be full with black squares. 🙈

Reviewer 2: ```suggestion
With this feature you can hide parts of your doc with a black rectangles. Opening a document without redaction is only possible if the user has permission for it, otherwise the page will be full with black squares. 🙈
```

<p align="center">
<img src="/img/posts/docviewer/redaction.gif">
</p>

**Highlight** ⭐

> The purpose of highlighting is to draw attention to important information in a text. Highlighting is effective because it asks the reader to check the important parts first, and then gives an effective way to review other later. 📑

Reviewer 2: ```suggestion
The purpose of highlighting is to draw attention to important information in a text. 📑
```

I think the first phrase is enough here.

<p align="center">
<img src="/img/posts/docviewer/highlight.gif">
</p>

**Annotation** 💬

> An annotation is extra information associated with a particular point in a document or other piece of information. It can be a note that includes a comment or explanation or anything what you would like to share with the audience in a designated place in the text. 📋

Reviewer 2: ```suggestion
Annotation is extra information associated with a particular point in a document. It can be a note that includes a comment, explanation or anything what you would like to share with the audience in a designated place in the text. 📋
```

<p align="center">
<img src="/img/posts/docviewer/annotation.gif">
</p>

**Usage**

> The operation of these three shapes is relatively the same with some differences:

Reviewer 2: ```suggestion
The operation of these three shapes is basically the same with some small differences:
```

- you can create them with a click-move-click method (check gifs) ✏️

> - you can delete them with `backspace` or `delete` keys (except annotations) and also if you do `right-click` on the shape and click on the delete button on the popout. 🚮

Reviewer 2: ```suggestion
- you can delete them with `backspace` or `delete` keys (except annotations) and also with a `right-click` on the shape then choosing delete button on the popout. 🚮
```

- you can also drag and drop the shapes if you would like to move them to another position on the page. ↖️↘️

> - right-clicking on shapes opens different popups, for the redactions and highlights only a delete button has appear, but for annotations a toolkit with different configuration options 🔧 where you can set line height, font size, font bold, font color and make your text to italic style.

Reviewer 2: ```suggestion
- right-clicking on shapes opens different popups. For the redactions and highlights only a delete button appears, but for annotations, a toolkit with different configuration options 🔧 shows up where you can set line height, font size, font color and make your text **bold** or _italic_.
```

<p align="center">
<img src="/img/posts/docviewer/right-click.png">
</p>

> # Whats next 📅

Reviewer 2: ```suggestion
# What's next 📅
```
